File: benchmarks/tsb/bench_cat_intersect_diff.ts
/**
 * Benchmark: catIntersectCategories / catDiffCategories — set operations on
 * categorical Series categories (100k-element Series with 20 categories each).
 * Outputs JSON: {"function": "cat_intersect_diff", "mean_ms": ..., "iterations": ..., "total_ms": ...}
 */
import { Series, catIntersectCategories, catDiffCategories } from "../../src/index.ts";

const SIZE = 100_000;
const WARMUP = 5;
const ITERATIONS = 30;

// Build two categorical Series with overlapping but not identical category sets
const catsA = Array.from({ length: 20 }, (_, i) => `cat_a_${i}`);
const catsB = Array.from({ length: 20 }, (_, i) => `cat_${i < 10 ? "a" : "b"}_${i}`);

const dataA = Array.from({ length: SIZE }, (_, i) => catsA[i % catsA.length]);
const dataB = Array.from({ length: SIZE }, (_, i) => catsB[i % catsB.length]);

const sA = new Series({ data: dataA }).cat.setCategories(catsA);
const sB = new Series({ data: dataB }).cat.setCategories(catsB);

for (let i = 0; i < WARMUP; i++) {
  catIntersectCategories(sA, sB);
  catDiffCategories(sA, sB);
}

const times: number[] = [];
for (let i = 0; i < ITERATIONS; i++) {
  const t0 = performance.now();
  catIntersectCategories(sA, sB);
  catDiffCategories(sA, sB);
  times.push(performance.now() - t0);
}

const total = times.reduce((a, b) => a + b, 0);
console.log(
  JSON.stringify({
    function: "cat_intersect_diff",
    mean_ms: total / ITERATIONS,
    iterations: ITERATIONS,
    total_ms: total,
  }),
);

File: benchmarks/pandas/bench_cat_intersect_diff.py
"""
Benchmark: pandas category set operations — intersection and difference of
categorical Series categories (100k-element, 20 categories each).
Mirrors tsb's catIntersectCategories / catDiffCategories.
Outputs JSON: {"function": "cat_intersect_diff", "mean_ms": ..., "iterations": ..., "total_ms": ...}
"""
import json
import time
import pandas as pd

SIZE = 100_000
WARMUP = 5
ITERATIONS = 30

cats_a = [f"cat_a_{i}" for i in range(20)]
cats_b = [f"cat_{'a' if i < 10 else 'b'}_{i}" for i in range(20)]

data_a = [cats_a[i % len(cats_a)] for i in range(SIZE)]
data_b = [cats_b[i % len(cats_b)] for i in range(SIZE)]

s_a = pd.Categorical(data_a, categories=cats_a)
s_b = pd.Categorical(data_b, categories=cats_b)

def cat_intersect(a, b):
    """Return new Categorical with categories = intersection of a.categories and b.categories."""
    b_set = set(b.categories)
    intersected = [c for c in a.categories if c in b_set]
    return pd.Categorical(a, categories=intersected)

def cat_diff(a, b):
    """Return new Categorical with categories = a.categories - b.categories."""
    b_set = set(b.categories)
    remaining = [c for c in a.categories if c not in b_set]
    return pd.Categorical(a, categories=remaining)

for _ in range(WARMUP):
    cat_intersect(s_a, s_b)
    cat_diff(s_a, s_b)

times = []
for _ in range(ITERATIONS):
    t0 = time.perf_counter()
    cat_intersect(s_a, s_b)
    cat_diff(s_a, s_b)
    times.append((time.perf_counter() - t0) * 1000)

total_ms = sum(times)
print(json.dumps({
    "function": "cat_intersect_diff",
    "mean_ms": total_ms / ITERATIONS,
    "iterations": ITERATIONS,
    "total_ms": total_ms,
}))
